source/extensions/filters/http/oauth2/oauth_client.h

```diff
@@ -90,7 +90,10 @@ class OAuth2ClientImpl : public OAuth2Client, Logger::Loggable<Logger::Id::upstr
   Http::RequestMessagePtr createPostRequest() {
     auto request = Http::Utility::prepareHeaders(uri_);
     request->headers().setReferenceMethod(Http::Headers::get().MethodValues.Post);
-    request->headers().setContentType(Http::Headers::get().ContentTypeValues.FormUrlEncoded);
+    request->headers().setReferenceContentType(
+        Http::Headers::get().ContentTypeValues.FormUrlEncoded);
+    request->headers().setReference(Http::CustomHeaders::get().Accept,
+                                    Http::Headers::get().ContentTypeValues.Json);
```

Comment:
Hmm do we want to make this configurable but on by default?

cc: @williamsfu99

Reply:
cc: @jparise

Reply:
As a note, the current client expects (to parse) JSON payload from the server: https://github.com/envoyproxy/envoy/blob/4d9fa702209a00cd1d9b407048af075d5c620435/source/extensions/filters/http/oauth2/oauth_client.cc#L81

Reply:
> Hmm do we want to make this configurable but on by default?

@rgs1 If Envoy is aiming for strict RFC 6749 compliance, I don't think it needs to be configurable. I'm not aware of a version of the OAuth 2.0 standard that allows `FormUrlEncoded` access token responses but I don't have much experience with OAuth to be honest. 

https://tools.ietf.org/html/rfc6749#section-4.1.4 states:
>  If the access token request is valid and authorized, the
   authorization server issues an access token and optional refresh
   token as described in Section 5.1.

https://tools.ietf.org/html/rfc6749#section-5.1 states:
> The authorization server issues an access token and optional refresh token, and constructs the response by adding the following parameters to the entity-body of the HTTP response with a 200 (OK) status code:
> [ ... ]
> The parameters are included in the entity-body of the HTTP response using the "application/json" media type as defined by [RFC4627].

I just noticed this also means Envoy's code was never really wrong to begin with. It's reasonable to expect a JSON response if the RFC says so, but of course setting the Accept header is nicer. GitHub just happened to violate the RFC by defaulting to `FormUrlEncoded` instead of JSON.

Reply:
Awesome, thanks for the reply. Mind adding a summary of the above with those links into the code as a comment?

Reply:
@rgs1 Done